src/transformers/adapters/peft_mixin.py

```diff
@@ -0,0 +1,217 @@
+
+        if not self._hf_peft_config_loaded:
+            self.peft_config = {}
+            self._hf_peft_config_loaded = True
+
+        adapter_config_file = find_adapter_config_file(
+            peft_model_id,
+            revision=revision,
+            use_auth_token=use_auth_token,
+            commit_hash=commit_hash,
+        )
+
+        if adapter_config_file is None:
+            raise ValueError(
+                f"adapter model file not found in {peft_model_id}. Make sure you are passing the correct path to the "
+                "adapter model."
+            )
+
+        loaded_peft_config = PeftConfig.from_pretrained(
+            peft_model_id,
+            revision=revision,
+            use_auth_token=use_auth_token,
+            commit_hash=commit_hash,
+        )
+
+        if not hasattr(loaded_peft_config, "target_modules"):
+            target_modules = TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING[self.config.model_type]
+            loaded_peft_config.target_modules = target_modules
+
+        if adapter_name not in self.peft_config:
+            self.peft_config[adapter_name] = loaded_peft_config
+        else:
+            raise ValueError(f"Adapter with name {adapter_name} already exists. Please use a different name.")
+
+        # Replace the adapter with the loaded adapter
+        create_and_replace(loaded_peft_config, self, adapter_name)
+
+        adapter_state_dict = load_peft_weights(
+            peft_model_id,
+            revision=revision,
+            use_auth_token=use_auth_token,
+        )
+
+        # We need to pre-process the state dict to remove unneeded prefixes - for backward compatibility
+        processed_adapter_state_dict = {}
+        for key, value in adapter_state_dict.items():
+            if "base_model.model" in key:
+                new_key = key.replace("base_model.model.", "")
+            else:
+                new_key = key
+            processed_adapter_state_dict[new_key] = value
+
+        # Load state dict
+        incompatible_keys = set_peft_model_state_dict(self, processed_adapter_state_dict, adapter_name)
+
+        if incompatible_keys is not None:
+            # check only for unexpected keys
+            if hasattr(incompatible_keys, "unexpected_keys") and len(incompatible_keys.unexpected_keys) > 0:
+                logger.warning(
+                    f"Loading adapter weights from {peft_model_id} led to unexpected keys not found in the model: "
+                    f" {incompatible_keys.unexpected_keys}. "
+                )
+
+        # @pacman100 why this was needed?
+        if (
+            (getattr(self, "hf_device_map", None) is not None)
+            and (len(set(self.hf_device_map.values()).intersection({"cpu", "disk"})) > 0)
+            and len(self.peft_config) == 1
+        ):
+            self._dispatch_accelerate_model(
+                device_map=device_map, max_memory=max_memory, offload_dir=offload_dir, offload_index=offload_index
```

Comment:
I think `PeftModel.from_pretrained(self, peft_model_id)` address all of this, WDYT?

Reply:
Yes correct ! However I preferred to not use `PeftModel` and have a granular control on each thing we do (inject adapter, get state_dict, load it, dispatch if cpu offload) so that it will be easier to maintain if in the future we add more features or some important changes in `PeftModel.from_pretrained`.

Reply:
But, on the contrary, if we fix `PeftModel.from_pretrained`, we need to update all the changes here too increasing the maintenance and breaking changes because `from_pretrained` would be always updated to be in sync with changes of `PeftModel` whereas here we might get an updated version of `PeftModel` but the current loading logic might not handle it. WDYT?